go/ippool/broker.go

```diff
@@ -0,0 +1,135 @@
+package ippool
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"time"
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+type ipAddressBroker struct {
+	manager    IPPoolManager
+	base       net.IP
+	startIndex int
+	endIndex   int
+	ownName    string
+	waitTime   time.Duration
+	ownIP      string
+	ownUsed    bool
+}
+
+// IPAddressBroker is the broker to retrieve a IP from the IP pool
+type IPAddressBroker = *ipAddressBroker
+
+var logName bool
+
+// NewIPAddressBroker creates a new instance
+func NewIPAddressBroker(manager IPPoolManager, base net.IP, startIndex, endIndex int, ownName string, waitTime time.Duration) (IPAddressBroker, error) {
+	return &ipAddressBroker{
+		manager:    manager,
+		base:       base,
+		startIndex: startIndex,
+		endIndex:   endIndex,
+		ownName:    ownName,
+		waitTime:   waitTime,
+	}, nil
+}
+
+func (b *ipAddressBroker) getExistingIPAddresses(ctx context.Context) (*IPPoolUsageLookupResult, error) {
+	return b.manager.UsageLookup(ctx, b.ownName)
+}
+
+func (b *ipAddressBroker) log(fmtstr string, args ...interface{}) {
+	if logName {
+		fmtstr = b.ownName + ": " + fmtstr
+	}
+	println(fmt.Sprintf(fmtstr, args...))
+}
+
+func (b *ipAddressBroker) announceIPAddress(ctx context.Context, used bool, lookupResult *IPPoolUsageLookupResult) error {
+	if lookupResult.OwnUsed {
+		return nil
+	}
+	var ip string
+	if used {
+		ip = lookupResult.OwnIP
+		if ip == "" {
+			return fmt.Errorf("internal error: own ip undefined")
+		}
+	} else {
+		ip = b.findFreeIPAddress(lookupResult)
+		if ip == "" {
+			return fmt.Errorf("no free IP address found")
+		}
+	}
+	if err := b.manager.SetIPAddress(ctx, b.ownName, ip, used); err != nil {
+		return err
+	}
+
+	b.ownIP = ip
+	b.ownUsed = used
+	return nil
+}
+
+func (b *ipAddressBroker) findFreeIPAddress(lookupResult *IPPoolUsageLookupResult) string {
+	for i := 0; i < 1000; i++ {
+		index := rand.Intn(b.endIndex-b.startIndex+1) + b.startIndex
+		base4 := b.base.To4()
+		ip := net.IPv4(base4[0], base4[1], base4[2], byte(index)).String()
```

Comment:
This will make the IPv6 folks unhappy...

Reply:
Well, the kube-apiserver makes IPv4 requests, therefore an IPv4 address is needed.
Are you suggesting to have an IPv6 network additionally?

Reply:
It depends how the kube-apiserver is configured. As of now, Gardener only supports IPv4, but kubernetes can also use IPv6 only or even dual-stack (IPv4 & IPv6). In an IPv6 only case this will not necessarily work, but we can park this for the time being.